File: bemullen_dharmesh/TransformMBTADwellTimes.py
# Filename: TransformMBTADwellTimes.py
# Author: Dharmesh Tarapore <dharmesh@bu.edu>
import urllib.request
from urllib.request import quote 
import json
import dml
import prov.model
import datetime
import uuid
import subprocess
import xmltodict

class TransformMBTADwellTimes(dml.Algorithm):
    contributor = "bemullen_dharmesh"
    reads = ["bemullen_dharmesh.mbta_red_dwells", "mbta_green_dwells"]
    writes = ["bemullen_dharmesh.mbta_red_dwells_monthly", "bemullen_dharmesh.mbta_green_dwells_monthly"]

    @staticmethod
    def execute(trial = False):
        startTime = datetime.datetime.now()

        # Set up the database connection.
        client = dml.pymongo.MongoClient()
        repo = client.repo
        repo.authenticate('bemullen_dharmesh', 'bemullen_dharmesh')

        subprocess.check_output('mongo repo -u bemullen_dharmesh -p\
            bemullen_dharmesh --authenticationDatabase "repo" transformMBTADwellTimes.js', shell=True)

        endTime = datetime.datetime.now()

        return {"start":startTime, "end":endTime}

    @staticmethod
    def provenance(doc = prov.model.ProvDocument(), startTime = None, endTime = None):
        client = dml.pymongo.MongoClient()
        repo = client.repo
        repo.authenticate('bemullen_dharmesh', 'bemullen_dharmesh')

        doc.add_namespace('alg', 'http://datamechanics.io/algorithm/') # The scripts are in <folder>#<filename> format.
        doc.add_namespace('dat', 'http://datamechanics.io/data/') # The data sets are in <user>#<collection> format.
        doc.add_namespace('ont', 'http://datamechanics.io/ontology#') # 'Extension', 'DataResource', 'DataSet', 'Retrieval', 'Query', or 'Computation'.
        doc.add_namespace('log', 'http://datamechanics.io/log/') # The event log.


        this_script = doc.agent('alg:bemullen_dharmesh#TransformCityScores', {prov.model.PROV_TYPE: prov.model.PROV['SoftwareAgent'], 'ont:Extension': 'py'})
        
        mbta_red_dwells = doc.entity('dat:bemullen_dharmesh#mbta_red_dwells',
            {prov.model.PROV_LABEL:'Dwell Intervals for the MBTA\'s Red Line', prov.model.PROV_TYPE:'ont:DataSet'})        
        get_mbta_red_dwells = doc.activity('log:uuid'+str(uuid.uuid4()), startTime, endTime,
            {'prov:label':'Dwell Intervals for the MBTA\'s Red Line'})
        doc.wasAssociatedWith(get_mbta_red_dwells, this_script)
        doc.used(get_mbta_red_dwells, mbta_red_dwells, startTime)
        doc.wasAttributedTo(mbta_red_dwells, this_script)
        doc.wasGeneratedBy(mbta_red_dwells, get_mbta_red_dwells, endTime) 

        mbta_green_dwells = doc.entity('dat:bemullen_dharmesh#mbta_red_dwells',
            {prov.model.PROV_LABEL:'Dwell Intervals for the MBTA\'s Green Line', prov.model.PROV_TYPE:'ont:DataSet'})        
        get_mbta_red_dwells = doc.activity('log:uuid'+str(uuid.uuid4()), startTime, endTime,
            {'prov:label':'Dwell Intervals for the MBTA\'s Green Line'})
        doc.wasAssociatedWith(get_mbta_green_dwells, this_script)
        doc.used(get_mbta_green_dwells, mbta_green_dwells, startTime)
        doc.wasAttributedTo(mbta_green_dwells, this_script)
        doc.wasGeneratedBy(mbta_green_dwells, get_mbta_green_dwells, endTime)        
        
        repo.logout()
        return doc

File: bemullen_dharmesh/transformMBTADwellTimes.js
/*
 * Filename: transformCodeEnforcements.js
 * Author: Dharmesh Tarapore <dharmesh@bu.edu>
 * Description: Data transformations on the MBTA dwell intervals dataset.
 */
d = db.bemullen_dharmesh;

mbta_red_dwells = d.mbta_red_dwells;
mbta_green_dwells = d.mbta_green_dwells;
// universities = d.universities;

// Create dates for MBTA data
mbta_red_dwells.find().forEach(function(doc) {
    dr = new Date(0);
    dr.setSeconds(parseFloat(doc.arr_dt));
    doc.date = dr;
    mbta_red_dwells.save(doc);
});

// Create dates for MBTA data
mbta_green_dwells.find().forEach(function(doc) {
    dr = new Date(0);
    dr.setSeconds(parseFloat(doc.arr_dt));
    doc.date = dr;
    mbta_green_dwells.save(doc);
});

mbta_red_dwells.find({dwell_time_sec: {$not: {$type: "int"}}}).forEach(function(doc) {
    doc.dwell_time_sec = parseInt(doc.dwell_time_sec)
    mbta_red_dwells.save(doc);
});

mbta_green_dwells.find({dwell_time_sec: {$not: {$type: "int"}}}).forEach(function(doc) {
    doc.dwell_time_sec = parseInt(doc.dwell_time_sec)
    mbta_green_dwells.save(doc);
});

var redBinned = mbta_red_binned = mbta_red_dwells.aggregate(
    [
    {
        $group: {
            _id: {
                month: {$month: "$date"},
                year: {$year: "$date"},
            },
            dwell_time: {$avg: "$dwell_time_sec"},
        },
    }]);

redArrayVals = redBinned.toArray();
d.mbta_red_dwells_monthly.remove({});
db.createCollection("mbta_red_dwells_monthly");
d.mbta_red_dwells_monthly.insertMany(redArrayVals);

var greenBinned = mbta_green_binned = mbta_green_dwells.aggregate(
    [
    {
        $group: {
            _id: {
                month: {$month: "$date"},
                year: {$year: "$date"},
            },
            dwell_time: {$avg: "$dwell_time_sec"},
        },
    }]);

greenArrayVals = greenBinned.toArray();
d.mbta_green_dwells_monthly.remove({});
db.createCollection("mbta_green_dwells_monthly");
d.mbta_green_dwells_monthly.insertMany(greenArrayVals);
